feat: support exit after single check

matthewadams opened the pull request:
This change allows the script to be scheduled via cron job or cloud scheduler.

matthewadams committed:
feat: support exit after single check

dbrand666 commented:
The thing about having an external scheduler trigger this script is that it's fundamentally a scraper application. If the auth token gets expired or invalidated or if the (undocumented) APIs change, the best that will happen is that the script will keep failing. Worst case, it keeps thinking the topup failed and tries to buy more data.

If we want the script to disable itself we need a tiny bit of persistence.

Maybe we should just return a fail exit code and leave it to the scheduler to deal with?

matthewadams committed:
chore: improve config typing, exit with nonzero on error

matthewadams commented:
Ok, I added exiting with an error code on any exception.  The problem is when `max_errors` is nonzero due to the program being written to support multiple pools.  I say only support one pool at a time, or throw if any pool update operation fails.  We can support `max_errors` tries on every rest call pretty easily with the following:

```
def with_tries(n: int, op):
    i = 0
    while True:
        try:
            return op()
        except:
            i += 1
            if i >= n:
                raise
```

I think `max_errors` would be reinterpreted as "max_tries - 1".  WDYT?

I also refactored config to be more typesafe up front.

dbrand666 commented on app/main.py:
```diff
@@ -99,27 +117,29 @@ def perform_topup(self) -> None:
             )
         else:
-            if bool(lifeguard.dryrun) is not False:
+            if configured_bool(lifeguard.dryrun) is True:
```
Is this cast still needed?

matthewadams replied:
oh no forgot that I handle it in `load_config` now -- my bad 😑

dbrand666 commented on app/main.py:
```diff
@@ -139,13 +159,25 @@ def perform_topup(self) -> None:
+            time.sleep(lifeguard.check_interval_minutes * 60)
+        sys.exit(0)
+    except Exception as e:
+        while True:
```
This could just be `traceback.print_exc(file=sys.stderr)`.

Shouldn't we be redirecting all logging to sys.stderr?

matthewadams replied:
If it's supposed to act like a command line app, then non-error logging should go to `stdout` & warning/error logging should go to `stderr`.  If it's supposed to act like a server app, then all logging should go to `stdout` period.

dbrand666 commented on app/main.py:
```diff
@@ -49,6 +64,9 @@ def poll(self) -> None:
             if pool.get_pool_data():
                 pool.perform_topup()
 
+        return self.config["check_interval_minutes"] > 0
```
I don't like this in here. It's not part of the polling logic.

matthewadams replied:
Refactor without your alternative?

matthewadams replied:
nvm -- 4ee4f80 addresses this

dbrand666 reviewed (changes requested):
You've come pretty close to convincing me this should just run once and exit ;-)

I really want it to disable itself at the first sign of problems though. How are you planning to handle that? Or aren't you concerned?

matthewadams committed:
chore: rm unnecessary usage of configured_bool

matthewadams committed:
chore: mv poll control logic to higher level

matthewadams commented:
WRT running once, I'd say remove the retry logic, unless you want to use my `with_tries` suggestion but only on rest calls.  Also, I'd remove the loop over all pools and just support a single pool per invocation.  Handling multiple pools could come later as required, IMHO.
